halo2_proofs/src/plonk/prover.rs

```diff
@@ -56,674 +33,38 @@ where
-        .into_iter()
-        .map(
-            |AdviceSingle {
-                 advice_polys,
-                 advice_blinds,
-             }| {
-                AdviceSingle {
-                    advice_polys: advice_polys
-                        .into_iter()
-                        .map(|poly| domain.lagrange_to_coeff(poly))
-                        .collect::<Vec<_>>(),
-                    advice_blinds,
-                }
-            },
-        )
+        .enumerate()
+        .map(|(i, circuit)| WitnessCalculator::new(params.k(), circuit, &config, &cs, instances[i]))
         .collect();
-
-    // Evaluate the h(X) polynomial
-    let h_poly = pk.ev.evaluate_h(
-        pk,
-        &advice
-            .iter()
-            .map(|a| a.advice_polys.as_slice())
-            .collect::<Vec<_>>(),
-        &instance
-            .iter()
-            .map(|i| i.instance_polys.as_slice())
-            .collect::<Vec<_>>(),
-        &challenges,
-        *y,
-        *beta,
-        *gamma,
-        *theta,
-        &lookups,
-        &shuffles,
-        &permutations,
-    );
-
-    // Construct the vanishing argument's h(X) commitments
-    let vanishing = vanishing.construct(params, domain, h_poly, &mut rng, transcript)?;
-
-    let x: ChallengeX<_> = transcript.squeeze_challenge_scalar();
-    let xn = x.pow([params.n()]);
-
-    if P::QUERY_INSTANCE {
-        // Compute and hash instance evals for each circuit instance
-        for instance in instance.iter() {
-            // Evaluate polynomials at omega^i x
-            let instance_evals: Vec<_> = meta
-                .instance_queries
-                .iter()
-                .map(|&(column, at)| {
-                    eval_polynomial(
-                        &instance.instance_polys[column.index()],
-                        domain.rotate_omega(*x, at),
-                    )
-                })
-                .collect();
-
-            // Hash each instance column evaluation
-            for eval in instance_evals.iter() {
-                transcript.write_scalar(*eval)?;
-            }
-        }
-    }
-
-    // Compute and hash advice evals for each circuit instance
-    for advice in advice.iter() {
-        // Evaluate polynomials at omega^i x
-        let advice_evals: Vec<_> = meta
-            .advice_queries
-            .iter()
-            .map(|&(column, at)| {
-                eval_polynomial(
-                    &advice.advice_polys[column.index()],
-                    domain.rotate_omega(*x, at),
-                )
-            })
-            .collect();
-
-        // Hash each advice column evaluation
-        for eval in advice_evals.iter() {
-            transcript.write_scalar(*eval)?;
-        }
+    let mut prover = ProverV2::<Scheme, P, _, _, _>::new(params, pk, instances, rng, transcript)?;
+    let mut challenges = HashMap::new();
+    let phases = prover.phases.clone();
+    for phase in &phases {
+        println!("DBG phase {}", phase.0);
```

Comment:
leftover

Reply:
Resolved in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/134143c618f41aea4805fd0cff14962c819397f1